File: asyncio/consumer.rb
require 'faraday'
require 'json'
require 'ksuid'
require 'nats/client'

class Iterfable
  URL = 'http://127.0.0.1:3000/email'.freeze

  def self.send(event)
    Faraday.post(URL, event.to_json, "Content-Type" => "application/json")
    print('*')
  end
end

NATS.start(:servers => ['nats://127.0.0.1:4222']) do
  puts 'Consuming pancakes...'
  NATS.subscribe('pancake.*') do |raw|
    print('.')
    event = JSON.parse(raw)
    Iterfable.send(event.merge(track_id: KSUID.new, ts: Time.now))
  end
end

File: asyncio/consumer.js
const chalk = require("chalk");
const fetch = require("node-fetch");
const NATS = require("nats");
const nats = NATS.connect({ servers: ["nats://127.0.0.1:4222"] });

process.on("SIGINT", () => {
  console.log("\nClosing connections...");
  nats.flush();
  nats.close();
  process.exit();
});

let msgId = 0;
const nextColor = () => {
  msgId = msgId + 5;
  return chalk.hsv(msgId % 360, 100, 100);
}

const Iterfable = {
  send: (event, color) => {
    fetch("http://127.0.0.1:3000/email", {
      method: "post",
      body: JSON.stringify(event),
      headers: { "Content-Type": "application/json" },
    }).finally(() => process.stdout.write(color.bold("●")));
  },
};

nats.subscribe(`pancake.*`, (text, reply, subject) => {
  const color = nextColor();
  process.stdout.write(color.bold("○"));
  const message = JSON.parse(text);
  Iterfable.send(message, color);
});
